platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/preconsensus/AsyncPreConsensusEventWriter.java

```diff
@@ -0,0 +1,190 @@
+     */
+    private final PreConsensusEventWriter writer;
+
+    /**
+     * Background work is performed on this thread.
+     */
+    private final MultiQueueThread handleThread;
+
+    /**
+     * Used to push the minimum generation non-ancient onto the handle queue.
+     */
+    private final BlockingQueueInserter<Long> minimumGenerationNonAncientInserter;
+
+    /**
+     * Used to push events onto the handle queue.
+     */
+    private final BlockingQueueInserter<EventImpl> eventInserter;
+
+    /**
+     * Create a new AsyncPreConsensusEventWriter.
+     *
+     * @param threadManager responsible for creating new threads
+     * @param config        preconsensus event stream configuration
+     * @param writer        the writer to which events will be written, wrapped by this class
+     */
+    public AsyncPreConsensusEventWriter(
+            final ThreadManager threadManager,
+            final PreConsensusEventStreamConfig config,
+            final PreConsensusEventWriter writer) {
+
+        this.writer = writer;
+
+        handleThread = new MultiQueueThreadConfiguration(threadManager)
+                .setComponent("pre-consensus")
+                .setThreadName("event-writer")
+                .setCapacity(config.writeQueueCapacity())
+                .addHandler(Long.class, this::setMinimumGenerationNonAncientHandler)
+                .addHandler(EventImpl.class, this::addEventHandler)
+                .build();
+
+        minimumGenerationNonAncientInserter = handleThread.getInserter(Long.class);
+        eventInserter = handleThread.getInserter(EventImpl.class);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void start() {
+        writer.start();
+        handleThread.start();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void stop() {
+        handleThread.stop();
+        writer.stop();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void writeEvent(final EventImpl event) throws InterruptedException {
+        if (event.getStreamSequenceNumber() == EventImpl.NO_STREAM_SEQUENCE_NUMBER
+                || event.getStreamSequenceNumber() == EventImpl.STALE_EVENT_STREAM_SEQUENCE_NUMBER) {
+            throw new IllegalStateException("Event must have a valid stream sequence number");
+        }
+        eventInserter.put(event);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setMinimumGenerationNonAncient(final long minimumGenerationNonAncient) throws InterruptedException {
+        minimumGenerationNonAncientInserter.put(minimumGenerationNonAncient);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void setMinimumGenerationToStore(final long minimumGenerationToStore) {
+        writer.setMinimumGenerationToStore(minimumGenerationToStore);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isEventDurable(final EventImpl event) {
```

Comment:
null annotations & checks + same for other methods / constructors in this PR

Reply:
Blocked on https://github.com/hashgraph/hedera-services/pull/5481